### ISSUE TITLE
Sometimes hangs

### ISSUE THREAD

**Issue by brettyj** [jishi/node-sonos-http-api#393]
Hi,

I love the app, but sometimes after a text to speech announcement the web request hangs and it doesn't restore the sonos state. Is there anyway I can turn on additional logging so I can have an idea of the issue?

Thanks

**Comment by brettyj** [comment]
Hi @jishi 

I think it's happened with both commands. It's not something I can recreate, I do find though that when I restart the sonos server api it'll work perfectly - I only seem to get the issue when the api has been running for a while.

I'll try running it with more verbose logging.

Thanks

**Comment by brettyj** [comment]
Also silly question, I've tried this:

```
set NODE_LOG_LEVEL=debug node server.js
npm start
```

and it's not working, am I doing something wrong? (I'm on Windows)

**Comment by jishi** [comment]
It might be that the subscriptions for events are not renewed correctly or something. It might be good to look in the Sonos logs as well, because they would tell you if the subscriptions have been terminated, and they would say for which IP they are terminated. 

It's normal to see termination for devices that goes on and off (basically, all other controllers), but you should never see it for a device that is always on (if you are running this on a server, or a raspberry pi, or similar). Also make sure that the device you are running this on, doesn't change IP frequently.

**Comment by jishi** [comment]
oh, you are on windows. Okay, then you need to do this:

`$env:NODE_LOG_LEVEL='debug'` dont forget the qoutes!
`node server.js` or `npm start`, doesn't matter.

**Comment by brettyj** [comment]
You sure that's right?

`C:\node-sonos-http-api>$env:NODE_LOG_LEVEL='debug'
The filename, directory name, or volume label syntax is incorrect.
`

Also forgive my ignorance but where can I find the sonos logs? I'm aware of the sonos web interface, is it on that?

Also the server that the API runs on is an always on server with a static ip.

**Comment by jishi** [comment]
It's powershell syntax, don't remember how you temporarily set environment variables in `cmd`...

Yes, it is in the sonos web interface, there is an entry called `/opt/anacapa.trace` or something similar under http://[IP of some player]:1400/status

**Comment by jishi** [comment]
`set NODE_LOG_LEVEL=debug`
`npm start`

might be what you want?

**Comment by brettyj** [comment]
Hi @jishi,

So I ran it a few times with debugging on and these are the logs:

```
WORKING
2016-12-01T13:58:42.423Z DEBUG pausing Bathroom
2016-12-01T13:58:42.424Z DEBUG invoking pause
2016-12-01T13:58:42.482Z DEBUG coordinator is Bathroom
2016-12-01T13:58:42.483Z DEBUG skipping grouping for Kitchen because it is already grouped with coordinator
2016-12-01T13:58:42.483Z DEBUG skipping grouping for Office because it is already grouped with coordinator
2016-12-01T13:58:42.484Z DEBUG setting volume 60 on Bathroom
2016-12-01T13:58:42.493Z DEBUG setting volume 60 on Kitchen
2016-12-01T13:58:42.502Z DEBUG setting volume 60 on Office
2016-12-01T13:58:42.584Z DEBUG repeat: false, shuffle: undefined
2016-12-01T13:58:42.584Z DEBUG calculated playmode to NORMAL
2016-12-01T13:58:42.693Z DEBUG invoking play
2016-12-01T13:58:42.816Z DEBUG Bathroom STOPPED
2016-12-01T13:58:43.019Z DEBUG Bathroom PLAYING
2016-12-01T13:58:45.356Z DEBUG Bathroom STOPPED
2016-12-01T13:58:45.357Z DEBUG announcement finished
2016-12-01T13:58:45.357Z DEBUG removing listener from Bathroom
2016-12-01T13:58:45.358Z DEBUG
  [ { players: [ [Object], [Object], [Object] ],
    state: 'PLAYING',
    uri: 'x-sonosapi-stream:s220687?sid=254&flags=8224&sn=0',
    metadata: '<DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp
/" xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/"><item id=
"-1" parentID="-1" restricted="true"><dc:title>LBC UK</dc:title><upnp:class>object.item.audioItem.audioBroadcast</upnp:c
lass><desc id="cdudn" nameSpace="urn:schemas-rinconnetworks-com:metadata-1-0/">SA_RINCON65031_</desc></item></DIDL-Lite>
',
    playMode: { repeat: 'none' } } ]
2016-12-01T13:58:45.360Z DEBUG coordinator is Bathroom
2016-12-01T13:58:45.361Z DEBUG skipping grouping for Kitchen because it is already grouped with coordinator
2016-12-01T13:58:45.361Z DEBUG skipping grouping for Office because it is already grouped with coordinator
2016-12-01T13:58:45.362Z DEBUG setting volume 32 on Bathroom
2016-12-01T13:58:45.368Z DEBUG setting volume 32 on Kitchen
2016-12-01T13:58:45.373Z DEBUG setting volume 32 on Office
2016-12-01T13:58:47.156Z DEBUG repeat: none, shuffle: undefined
2016-12-01T13:58:47.157Z DEBUG calculated playmode to NORMAL
2016-12-01T13:58:47.267Z DEBUG invoking play


WORKED, BUT VERY SLOW (look at the timestamps):
2016-12-01T13:58:54.792Z DEBUG pausing Bathroom
2016-12-01T13:58:54.793Z DEBUG invoking pause
2016-12-01T13:58:54.835Z DEBUG coordinator is Bathroom
2016-12-01T13:58:54.835Z DEBUG skipping grouping for Kitchen because it is already grouped with coordinator
2016-12-01T13:58:54.835Z DEBUG skipping grouping for Office because it is already grouped with coordinator
2016-12-01T13:58:54.836Z DEBUG setting volume 60 on Bathroom
2016-12-01T13:58:54.843Z DEBUG setting volume 60 on Kitchen
2016-12-01T13:58:54.846Z DEBUG setting volume 60 on Office
2016-12-01T13:58:54.865Z DEBUG No such service x-sonosapi-stream:s220687?sid=254&flags=8224&sn=0
2016-12-01T13:58:54.936Z DEBUG repeat: false, shuffle: undefined
2016-12-01T13:58:54.937Z DEBUG calculated playmode to NORMAL
2016-12-01T13:58:55.042Z DEBUG invoking play
2016-12-01T13:58:58.982Z DEBUG Bathroom STOPPED
2016-12-01T14:00:54.798Z DEBUG pausing Bathroom
2016-12-01T14:00:54.798Z DEBUG invoking pause
2016-12-01T14:00:54.806Z DEBUG coordinator is Bathroom
2016-12-01T14:00:54.807Z DEBUG skipping grouping for Kitchen because it is already grouped with coordinator
2016-12-01T14:00:54.807Z DEBUG skipping grouping for Office because it is already grouped with coordinator
2016-12-01T14:00:54.807Z DEBUG setting volume 60 on Bathroom
2016-12-01T14:00:54.812Z DEBUG setting volume 60 on Kitchen
2016-12-01T14:00:54.817Z DEBUG setting volume 60 on Office
2016-12-01T14:00:54.935Z DEBUG repeat: false, shuffle: undefined
2016-12-01T14:00:54.935Z DEBUG calculated playmode to NORMAL
2016-12-01T14:00:55.043Z DEBUG invoking play
2016-12-01T14:00:55.350Z DEBUG Bathroom PLAYING
2016-12-01T14:00:55.350Z DEBUG Bathroom PLAYING
2016-12-01T14:00:57.712Z DEBUG Bathroom STOPPED
2016-12-01T14:00:57.712Z DEBUG announcement finished
2016-12-01T14:00:57.713Z DEBUG Bathroom STOPPED
2016-12-01T14:00:57.713Z DEBUG announcement finished
2016-12-01T14:00:57.714Z DEBUG removing listener from Bathroom
2016-12-01T14:00:57.714Z DEBUG removing listener from Bathroom
2016-12-01T14:00:57.715Z DEBUG
  [ { players: [ [Object], [Object], [Object] ],
    state: 'PLAYING',
    uri: 'x-sonosapi-stream:s220687?sid=254&flags=8224&sn=0',
    metadata: '<DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp
/" xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/"><item id=
"-1" parentID="-1" restricted="true"><dc:title>LBC UK</dc:title><upnp:class>object.item.audioItem.audioBroadcast</upnp:c
lass><desc id="cdudn" nameSpace="urn:schemas-rinconnetworks-com:metadata-1-0/">SA_RINCON65031_</desc></item></DIDL-Lite>
',
    playMode: { repeat: 'none' } } ]
2016-12-01T14:00:57.717Z DEBUG
  [ { players: [ [Object], [Object], [Object] ],
    state: 'STOPPED',
    uri: 'TTLURL',
    metadata: '',
    playMode: { repeat: 'none' },
    trackNo: 1,
    elapsedTime: 0 } ]
2016-12-01T14:00:57.720Z DEBUG coordinator is Bathroom
2016-12-01T14:00:57.720Z DEBUG skipping grouping for Kitchen because it is already grouped with coordinator
2016-12-01T14:00:57.720Z DEBUG skipping grouping for Office because it is already grouped with coordinator
2016-12-01T14:00:57.721Z DEBUG coordinator is Bathroom
2016-12-01T14:00:57.722Z DEBUG skipping grouping for Kitchen because it is already grouped with coordinator
2016-12-01T14:00:57.722Z DEBUG skipping grouping for Office because it is already grouped with coordinator
2016-12-01T14:00:57.723Z DEBUG setting volume 32 on Bathroom
2016-12-01T14:00:57.724Z DEBUG setting volume 60 on Bathroom
2016-12-01T14:00:57.730Z DEBUG setting volume 60 on Kitchen
2016-12-01T14:00:57.733Z DEBUG setting volume 60 on Office
2016-12-01T14:00:57.735Z DEBUG setting volume 32 on Kitchen
2016-12-01T14:00:57.738Z DEBUG setting volume 32 on Office
2016-12-01T14:00:57.785Z DEBUG repeat: none, shuffle: undefined
2016-12-01T14:00:57.786Z DEBUG calculated playmode to NORMAL
2016-12-01T14:00:59.577Z DEBUG repeat: none, shuffle: undefined
2016-12-01T14:00:59.578Z DEBUG calculated playmode to NORMAL
2016-12-01T14:00:59.793Z DEBUG invoking play
2016-12-01T14:00:59.798Z ERROR Error
    at Object.invoke (C:\node-sonos-http-api\node_modules\sonos-discovery\lib\helpers\soap.js:95:10)
    at Player.trackSeek (C:\node-sonos-http-api\node_modules\sonos-discovery\lib\models\Player.js:469:15)
    at C:\node-sonos-http-api\node_modules\sonos-discovery\lib\prototypes\SonosSystem\applyPreset.js:124:46
    at process._tickCallback (node.js:369:9)
2016-12-01T14:01:00.361Z DEBUG No such service x-sonosapi-stream:s220687?sid=254&flags=8224&sn=0
```

Does that help? It seems to get stuck on Bathroom STOPPED. Nothing obvious was in the anacapa.trace file.

**Comment by jishi** [comment]
Yeah, it seems like it has retried the sayall command 2 extra times, because the first one hanged at 
`2016-12-01T13:58:58.982Z DEBUG Bathroom STOPPED`

It would expect a PLAYING row, which is an indication that the announcment started playing, but that never arrives. After two minutes, it seemed to have retried the request (did you call this in a browser?), and for some odd reason the events seems to have been duplicated the second time around. 

It at least looks consistent to my findings, I'll need to figure out if this is solvable or not. At least try to add the timeout functionality for sayall.

**Comment by brettyj** [comment]
Yeah I was testing this in a browser - it's odd, it'll work 90% of the time, then fail randomly. It seems to be getting the state changes fine from sonos though (as it knows it's changed to STOPPED). If there's anything I can do to help debug this, I'm more than willing to help

**Comment by jishi** [comment]
If you find any combination of phrase that consistently fails, it would be interesting to know.

**Comment by brettyj** [comment]
It seems to work 90% of the time, then get stuck - I've tried with wget as well to rule out browser weirdness. It always gets stuck on that bathroom STOPPED line

**Comment by jishi** [comment]
I'm evaluating a few changes that might make it a bit more robust. It's a bit tedious to test since it is really hard to reproduce consistently though. 

Would you have the possibility to test a branch easily to see if you can notice an improvement? If that is the case, I'll consider pushing the changes I have for you to test, since you would have a reference on how often it fails.

**Comment by brettyj** [comment]
Hi @jishi 

I'm currently going through a house move and haven't got proper internet installed yet, due to rubbish BT, however as soon as that's done, I'm more than willing to test a branch for you!

Thanks

**Comment by jishi** [comment]
I have merged the improvements I've made in 1.2.7. Test it out and re-open if you can still reproduce this.
